packages/gatsby-dev-cli/src/local-npm-registry/install-packages.js

```diff
@@ -1,22 +1,82 @@
+const path = require(`path`)
+const fs = require(`fs-extra`)
+
 const { promisifiedSpawn } = require(`../utils/promisified-spawn`)
 const { registryUrl } = require(`./verdaccio-config`)
 
-const installPackages = async ({ packagesToInstall }) => {
-  const installCmd = [
-    `yarn`,
-    [
-      `add`,
-      ...packagesToInstall.map(packageName => `${packageName}@gatsby-dev`),
-      `--registry=${registryUrl}`,
-      `--exact`,
-    ],
-  ]
-
+const installPackages = async ({ packagesToInstall, yarnWorkspaceRoot }) => {
   console.log(
     `Installing packages from local registry:\n${packagesToInstall
       .map(packageAndVersion => ` - ${packageAndVersion}`)
       .join(`\n`)}`
   )
+  let installCmd
+  if (yarnWorkspaceRoot) {
+    // this is very hacky - given root, we run `yarn workspaces info`
+    // to get list of all workspaces and their locations, and manually
+    // edit package.json file for packages we want to install
+    // to make sure there are no mismatched versions of same package
+    // in workspaces which should preserve node_modules structure
+    // (packages being mostly hoisted to top-level node_modules)
+
+    const { stdout } = await promisifiedSpawn([
+      `yarn`,
+      [`workspaces`, `info`, `--json`],
+      { stdio: `pipe` },
+    ])
+
+    const workspacesLayout = JSON.parse(JSON.parse(stdout).data)
+
+    const handleDeps = deps => {
+      if (!deps) {
+        return false
+      }
+
+      let changed = false
+      Object.keys(deps).forEach(depName => {
+        if (packagesToInstall.includes(depName)) {
+          deps[depName] = `gatsby-dev`
+          changed = true
+        }
+      })
+      return changed
+    }
+
+    Object.keys(workspacesLayout).forEach(workspaceName => {
+      const { location } = workspacesLayout[workspaceName]
+      const pkgJsonPath = path.join(yarnWorkspaceRoot, location, `package.json`)
+      if (!fs.existsSync(pkgJsonPath)) {
+        return
+      }
+      const pkg = JSON.parse(fs.readFileSync(pkgJsonPath, `utf8`))
+
+      let changed = false
+      changed |= handleDeps(pkg.dependencies)
```

Comment:
Nice Bitwise OR!

But I'd change this to 
```
changed = changed || handleDeps(pkg.dependencies) || handleDeps(pkg.devDependencies) || handleDeps(pkg.peerDependencies)
```
for readability

Reply:
This will change behaviour - if `handleDeps(pkg.dependencies)` then next one won't ever be called and I do need to change it everywhere

Reply:
So to avoid `|=` thsi would need to be:
```
if (handleDeps(pkg.dependencies)) {
  changed = true
}
if (handleDeps(pkg.devDependencies)) {
  changed = true
}
if (handleDeps(pkg.peerDependencies)) {
  changed = true
}
```

Reply:
Ah, I'd typically want to avoid functions with side effects in conditionals due to this short circuiting 

But not a blocker for now!